rust/kernel/pm_runtime.rs

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Device run-time power management helper functions.
+//!
+//! C header: [`include/linux/pm_runtime.h`](../../../../include/linux/pm_runtime.h)
+
+use crate::{
+    bindings,
+    device::{self, RawDevice},
+    error::Result,
+    to_result,
+};
+
+/// Enable runtime PM of device `dev`.
+pub fn enable(dev: &impl RawDevice) -> DisableGuard {
+    // SAFETY: `dev` is valid by the type invariant.
+    unsafe { bindings::pm_runtime_enable(dev.raw_device()) };
+    DisableGuard::new(device::Device::from_dev(dev))
+}
+
+/// Guard that disables runtime PM of device `dev`.
+pub struct DisableGuard {
+    dev: device::Device,
+}
+
+impl DisableGuard {
+    /// Create new instance of a guard of device `dev`.
+    pub fn new(dev: device::Device) -> Self {
+        Self { dev }
+    }
+}
+
+impl Drop for DisableGuard {
+    fn drop(&mut self) {
+        // SAFETY: `dev` is valid by the type invariant.
+        unsafe { bindings::pm_runtime_disable(self.dev.raw_device()) };
+    }
+}
+
+/// Drop runtime PM usage counter of a device.
+/// Decrement the runtime PM usage counter of `dev` unless it is 0 already.
+pub fn put_noidle(dev: &impl RawDevice) {
+    // SAFETY: `dev` is valid by the type invariant.
+    unsafe { bindings::pm_runtime_put_noidle(dev.raw_device()) };
```

Comment:
If this causes the device to be suspended will operations on the device error out or will it result in UB?

Reply:
Say the PM usage counter is 1 and some other code increments it and later decrements it using one of the other methods which does have side effects. If you were to run `pm_runtime_put_noidle` between the increment and the decrement, the decrement would see 0. Can the side effects caused by the other decrement cause UB?

Pseudo code:

```c
pm_runtime_resume_and_get(dev);
// do some work
pm_runtime_put_autosuspend(dev);
```

Can there be UB if another task does

```c
pm_runtime_put_noidle(dev)
```

between the get and put of the first task?

Reply:
Hi,

First of all, I would like to apologize for the rather late reply.
The scenario you provided is not a UB Rust's point of view. It will cause a counter underflow and possible device misbehavior. There have already been some changes in the topic of handling this case, specifically:
[PM: runtime: Avoid device usage count underflows](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/drivers/base/power/runtime.c?h=next-20220504&id=82586a721595fde2c1bc0dac7deed2fe9d42c106)